### ISSUE TITLE
Assertion error in Singular in gamma_normalize

### ISSUE THREAD

**Issue by embray** [pynac/pynac#244]
As originally mentioned [here](https://trac.sagemath.org/timeline?from=2017-04-18T13%3A18%3A47Z&precision=second) when Singular is built in debug mode, and is thus configured to abort when an assertion fails, there is an error that can occur while doing polynomial GCD with Singular + flint.  I was able to simplify this to the following test case:

```
sage: var('n')
n
sage: ((n - 1)^2*gamma(2*n + 5)/gamma(n + 3) + gamma(2*n + 3)/gamma(n + 1)).gamm
....: a_normalize()
error: non-immediate integer expected
canonicalform.cc:53: In function `void CanonicalForm::mpzval(__mpz_struct*) const':
failed assertion `!is_imm (value) && value->levelcoeff() == IntegerDomain'

*** SIG 6 *** inside sig_on
do_raise_exception(sig=6)
PyErr_Occurred() = (nil)
Raising Python exception 0 ms after signals...
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-7adf30876b61> in <module>()
----> 1 ((n - Integer(1))**Integer(2)*gamma(Integer(2)*n + Integer(5))/gamma(n + Integer(3)) + gamma(Integer(2)*n + Integer(3))/gamma(n + Integer(1))).gamma_normalize()

/home/embray/src/sagemath/sage/src/sage/symbolic/expression.pyx in sage.symbolic.expression.Expression.gamma_normalize (build/cythonized/sage/symbolic/expression.cpp:52670)()
   9686         """
   9687         cdef GEx x
-> 9688         sig_on()
   9689         try:
   9690             x = gamma_normalize(self._gobj)

src/cysignals/signals.pyx in cysignals.signals.sig_raise_exception (build/src/cysignals/signals.c:1475)()

RuntimeError: Aborted
```

The relevant portion of the C-level stack trace is:

```
#0  __GI_abort () at abort.c:51
#1  0x00007f7330c21062 in CanonicalForm::mpzval (this=0x7fffeb072a10,
    val=0x7fffeb0729e0) at canonicalform.cc:53
#2  0x00007f7330da28a1 in convertCF2Fmpz (result=0x3869bb0, f=...)
    at FLINTconvert.cc:68
#3  0x00007f7330da29a3 in convertFacCF2Fmpz_poly_t (result=0x7fffeb072a90,
    f=...) at FLINTconvert.cc:79
#4  0x00007f7330cb6358 in gcd_univar_flint0 (F=..., G=...) at cfUnivarGcd.cc:63
#5  0x00007f7330cb5788 in subResGCD_0 (f=..., g=...) at cfSubResGcd.cc:187
#6  0x00007f7330c5abf8 in gcd_poly (f=..., g=...) at cf_gcd.cc:138
#7  0x00007f7330c5bc2e in gcd (f=..., g=...) at cf_gcd.cc:321
#8  0x00007f732a467581 in GiNaC::gcdpoly (a=..., b=...,
    ca=ca@entry=0x7fffeb073480, cb=cb@entry=0x7fffeb073490,
    check_args=check_args@entry=false) at mpoly-singular.cpp:498
#9  0x00007f732a468638 in GiNaC::gcdpoly (a=..., b=...,
    ca=ca@entry=0x7fffeb0738f0, cb=cb@entry=0x7fffeb073950,
    check_args=check_args@entry=false) at mpoly-singular.cpp:443
#10 0x00007f732a467d59 in GiNaC::gcdpoly (a=..., b=...,
    ca=ca@entry=0x7fffeb073ce0, cb=cb@entry=0x7fffeb073d40,
    check_args=check_args@entry=false) at mpoly-singular.cpp:373
#11 0x00007f732a467979 in GiNaC::gcdpoly (a=..., b=...,
    ca=ca@entry=0x7fffeb073f50, cb=cb@entry=0x7fffeb073f60,
    check_args=check_args@entry=false) at mpoly-singular.cpp:354
#12 0x00007f732a483e82 in GiNaC::frac_cancel (n=..., d=...) at normal.cpp:405
#13 0x00007f732a48503b in GiNaC::mul::normal (this=<optimized out>,
    repl=std::map with 2 elements, rev_lookup=std::map with 2 elements,
    level=<optimized out>, options=<optimized out>) at normal.cpp:522
#14 0x00007f732a4845b6 in GiNaC::add::normal (this=0x3c7b6b0,
    repl=std::map with 2 elements, rev_lookup=std::map with 2 elements,
    level=0, options=0) at normal.cpp:450
#15 0x00007f732a47e7c5 in GiNaC::ex::numer_denom (
    this=this@entry=0x7fffeb074870) at normal.cpp:686
#16 0x00007f732a47e90a in GiNaC::factor (the_ex=..., res_ex=...)
    at normal.cpp:1045
#17 0x00007f732a4d61c8 in GiNaC::gamma_normalize (the_ex=...) at sum.cpp:326
```
https://trac.sagemath.org/ticket/22826

**Comment by rwst** [comment]
Minimal test case:
```
sage: (n-QQ(1)).gcd(n+QQ(1))
```

**Comment by embray** [comment]
Ah, that makes sense.  The thing I don't understand about the test case I had is where exactly the rationals were coming in.

**Comment by rwst** [comment]
It appears Singular's authors don't want it to be fed with rationals that have denominator `1` (so make it an integer) although the GCD seems to work nevertheless. Fixed via 15cfffd8d8fd03aacc53839f77893ee2e364c8af
